rmw_connextdds_common/src/ndds/dds_api_ndds.cpp

```diff
@@ -1275,3 +1310,113 @@ rmw_connextdds_enable_security(
+  auto scope_exit_cft_parameters = rcpputils::make_scope_exit(
+    [&cft_parameters]() {
+      if (!DDS_StringSeq_finalize(&cft_parameters)) {
+        RMW_CONNEXT_LOG_ERROR_SET("failed to finalize cft parameters sequence")
+      }
+    });
+  if (expression_parameters) {
+    if (RMW_RET_OK != initialize_cft_parameters(&cft_parameters, expression_parameters)) {
+      RMW_CONNEXT_LOG_ERROR_SET("failed to initialize_cft_parameters")
+      return RMW_RET_ERROR;
+    }
+  }
+
+  DDS_ReturnCode_t ret =
+    DDS_ContentFilteredTopic_set_expression(cft_topic, filter_expression, &cft_parameters);
+  if (DDS_RETCODE_OK != ret) {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to set content-filtered topic")
+    return RMW_RET_ERROR;
+  }
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+rmw_connextdds_get_cft_filter_expression(
+  DDS_TopicDescription * const topic_desc,
+  char ** const expr_out,
+  rcutils_string_array_t * cft_params_out)
+{
+  DDS_ContentFilteredTopic * const cft_topic =
+    DDS_ContentFilteredTopic_narrow(topic_desc);
+
+  int parameters_len;
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  // get filter_expression
+  const char * expression = DDS_ContentFilteredTopic_get_filter_expression(cft_topic);
+  if (!expression) {
+    RMW_SET_ERROR_MSG("failed to get filter expression");
+    return RMW_RET_ERROR;
+  }
+
+  *expr_out = rcutils_strdup(expression, allocator);
+  if (NULL == *expr_out) {
+    RMW_SET_ERROR_MSG("failed to duplicate string");
+    return RMW_RET_BAD_ALLOC;
+  }
+  auto scope_exit_filter_expression_delete =
+    rcpputils::make_scope_exit(
+    [expr_out, allocator]()
+    {
+      if (*expr_out) {
+        allocator.deallocate(*expr_out, allocator.state);
+        *expr_out = nullptr;
+      }
+    });
+
+  // get parameters
+  struct DDS_StringSeq parameters;
+  DDS_ReturnCode_t status =
+    DDS_ContentFilteredTopic_get_expression_parameters(cft_topic, &parameters);
+  if (DDS_RETCODE_OK != status) {
+    RMW_SET_ERROR_MSG("failed to get expression parameters");
+    return RMW_RET_ERROR;
+  }
+  auto scope_exit_parameters_delete =
+    rcpputils::make_scope_exit(
+    [&parameters]()
+    {
+      DDS_StringSeq_finalize(&parameters);
+    });
+
+  parameters_len = DDS_StringSeq_get_length(&parameters);
+  rcutils_ret_t rcutils_ret =
+    rcutils_string_array_init(cft_params_out, parameters_len, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG("failed to init string array for expression parameters");
+    return RMW_RET_ERROR;
+  }
+  auto scope_exit_expression_parameters_delete =
+    rcpputils::make_scope_exit(
+    [cft_params_out]()
+    {
+      if (RCUTILS_RET_OK != rcutils_string_array_fini(cft_params_out)) {
+        RCUTILS_LOG_ERROR("Error while finalizing expression parameter due to another error");
+      }
+    });
+  for (int i = 0; i < parameters_len; ++i) {
+    char * parameter = rcutils_strdup(DDS_StringSeq_get(&parameters, i), allocator);
```

Comment:
Please put call to `DDS_StringSeq_get()` on a separate line and cache result to a local variable. Also, although they are equivalent, please use `*DDS_StringSeq_get_reference(&parameters, i)`, since the `get_reference()` function is used for all sequence access in the rest of the RMW. Thanks!

Reply:
updated